### PR TITLE
Events IIc (supplemental) - Return of the Slack

### PR DESCRIPTION
**Pull request by SolarDrew**
# Description

Second attempt at #1071 after far too long letting it go stale and die.

Follows from PRs #951 and #1069. Adds Slack-specific event types and appropriate handling by the connector.

## Status
**UNDER DEVELOPMENT**

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# How Has This Been Tested?

I'm in the process of adding tests to the test suite.

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

Fixes SolarDrew/skill-picard#30

### CODE REVIEW

File: opsdroid/connector/slack/__init__.py [hunk 1]
```diff
@@ -153,7 +155,7 @@ def __init__(self, config, opsdroid=None):
         message["text"] = await self.replace_usernames(message["text"])
 
         await self.opsdroid.parse(
```
Comment (Cadair) [inline opsdroid/connector/slack/__init__.py]: this should call the new event creator.

File: opsdroid/connector/slack/__init__.py [hunk 2]
```diff
@@ -309,3 +311,32 @@ def __init__(self, config, opsdroid=None):
                 )
 
         return aiohttp.web.Response(text=json.dumps("Received"), status=200)
+
+    @register_event(opsdroid.events.NewRoom)
+    async def _send_room_creation(self, creation_event):
+        params = creation_event.room_params
+        params = params.get("slack", params)
```
Comment (Cadair) [inline opsdroid/connector/slack/__init__.py]: you aren't doing anything with these params, and didn't we drop this anyway?

File: opsdroid/connector/slack/__init__.py [hunk 3]
```diff
@@ -309,3 +311,32 @@ def __init__(self, config, opsdroid=None):
...
+        )
+
+    @register_event(opsdroid.events.JoinRoom)
+    async def _send_join_room(self, join_event):
+        return await self.slacker.channels.join(join_event.target, "true")
+
+    @register_event(opsdroid.events.UserInvite)
+    async def _send_user_invitation(self, invite_event):
+        return await self.slacker.channels.invite(
+            invite_event.target, invite_event.user
```
Comment (Cadair) [inline opsdroid/connector/slack/__init__.py]:
```suggestion
            invite_event.target, invite_event.user_id
```

File: opsdroid/connector/slack/__init__.py (hunk 2 above)
Comment (SolarDrew) [reply opsdroid/connector/slack/__init__.py]: Yep, think that's a bad merge :+1:

**Comment by Cadair** [comment]
bad stale bot.

**Comment by SolarDrew** [comment]
This is pretty much ready for review, except that I don't understand the one test failure I'm still getting. The event creator wrapper is telling me that the creator's `connector.opsdroid` is `None`. But it's only doing this on whichever test runs first from `TestEventCreatorAsync`, and running it again with `--lf` makes the failure go away, so it's clearly something to do with the state or setup of that test class but I have no idea how to figure out what.

**Comment by Cadair** [comment]
The travis build passed: https://travis-ci.org/github/opsdroid/opsdroid/builds/681507774 It just didn't tell us about it.

**Pull request merged**